Roster refactor

vonbearshark opened the pull request:
## Background

I link to relevant tickets, but the changes have broader implications, as I ended up touching most of the Roster code.

## GitHub Issue
#1265 
#1187 

## Associated PRs
N/A

## Validation Plan
- [x] Previous month active filter works, displays the correct children (#1187)
- [x] Withdrawn filter works, displays correct children
- [x] Withdrawn and month filters are mutually exclusive (clicking `withdrawn` from month resets state to `withdrawn`)
- [x] Site filtering still works as expected (including for multiorg users)
- [x] Multiorg filtering works as expected
- [x] For multiorg users:  First load of roster filters the initial `All Sites` to the appropriate org's data
- [x] Batch editing works?
- [x] Back buttons work in the Roster view and Child views (create, edit, etc.)

## Automated Testing
- [x] All unit tests are passing.
- [x] All e2e tests are passing.
- [x] If applicable, unit tests have been ~added~ updated to cover this changeset.
- [x] If applicable, e2e tests have been ~added~ updated to cover this changeset.

vonbearshark commented:
We can _probably_ review merge these changes independent of the `TODO`s (without pagination, filtering, sql-izing), making this just an intermediary refactor step. Or by @jlhogan 's suggestion we could keep this as a feature branch and add those features in

vonbearshark commented:
@JordanGuinn I tried to cover all the changes from https://github.com/ctoec/data-collection/pull/1136/files, but lmk what I missed

vonbearshark commented on client/src/containers/App/App.tsx:
```diff
@@ -5,7 +5,6 @@ import { ErrorBoundary, Header } from '@ctoec/component-library';
 import routes from '../../routes';
 import MakeRouteWithSubRoutes from './MakeRouteWithSubroute';
 import UserContext from '../../contexts/UserContext/UserContext';
-import { mailToLinkProps } from '../../components/MailToLink';
```
unrelated, unsued var

vonbearshark commented on src/utils/processChild.ts:
```diff
@@ -78,3 +65,97 @@ function distributeValidationErrorsToSubObjects<
+  if (child.enrollments) {
+    child.enrollments = child.enrollments.sort((enrollmentA, enrollmentB) => {
+      if (enrollmentA.fundings) {
+        enrollmentA.fundings = enrollmentA.fundings.sort((fundingA, fundingB) =>
+          propertyDateSorter(
+            fundingA,
+            fundingB,
+            (f) => f.firstReportingPeriod.period
+          )
+        );
+      }
+
+      return propertyDateSorter(enrollmentA, enrollmentB, (e) => e.entry);
+    });
+  }
+
+  if (child.family?.incomeDeterminations) {
+    child.family.incomeDeterminations = child.family.incomeDeterminations.sort(
+      (determinationA, determinationB) =>
+        propertyDateSorter(
+          determinationA,
+          determinationB,
+          (d) => d.determinationDate
+        )
+    );
+  }
+};
+
+/**
+ * Apply all post-processing to a child record:
+ * 	- remove deleted entities
+ * 	- sort entities by date
+ * 	- validate full object tree
+ * @param child
+ */
+export const postProcessChild = async (child: Child) => {
+  removeDeletedEntitiesFromChild(child);
+  sortEntities(child);
+  return validateObject(child);
+};
```
Consolidated these helper fns here, otherwise untouched

vonbearshark commented on client/src/containers/Home/PostSubmit.tsx:
```diff
@@ -34,33 +34,26 @@ export const PostSubmitHome: React.FC = () => {
-      .catch((err) => {
-        throw new Error(err);
-      });
+    apiGet(`children/metadata?showFundings=${showFundings}`, accessToken)
```
was able to consolidate these calls into the `/metadata` endpoint, where showing funding is optional

vonbearshark replied:
I am not sure I see the difference. In the previous code we would make the two other calls and conditionally make the funding call. Here we are making one call (in place of the previous two), and passing this conditional param, which optionally adds the controller call on the backend.right?

vonbearshark replied:
(possible I might have edited this afterward, losing track at this point)

vonbearshark commented on client/src/containers/Home/PreSubmit.tsx:
```diff
@@ -38,10 +38,8 @@ export const PreSubmitHome: React.FC = () => {
-        .then((res) => {
-          setFundingSpacesDisplay(res.fundingSpacesMap);
-        })
+      apiGet('funding-spaces/funding-map', accessToken)
```
funding map also has it's own endpoint, which uses the same controller fn as the `metadata` endpoint

vonbearshark commented on src/controllers/enrollmentReports/formatUploadPreview.ts:
```diff
@@ -5,7 +5,6 @@ import {
   getLastEnrollment,
   getLastFunding,
 } from '../../../client/src/utils/models';
-import { validate } from 'class-validator';
```
unrelated, unused var

vonbearshark commented on client/src/contexts/RosterContext/RosterContext.tsx:
```diff
@@ -1,39 +1,98 @@
-import React, { useState } from 'react';
+import React, { useContext, useEffect } from 'react';
```
Refactored this to handle the roster data: the query and the child records. Now one course of truth for all child/sibling components

vonbearshark commented on client/src/contexts/RosterContext/RosterContext.tsx:
```diff
@@ -91,15 +150,37 @@ const RosterProvider: React.FC = ({ children }) => {
         updateCurrentRosterCache,
-        setRosterQuery,
-        rosterQuery,
+        // Function to update active month, to pass down into month filter buttons
+        updateActiveMonthQuery: (newMonth?: Moment) => {
+          const month = getQueryMonthFormat(newMonth);
+          history.push({
+            search: stringify({
+              ...query,
+              withdrawn: defaultCtx.query.withdrawn,
+              month,
+            }),
+          });
+        },
+        // Function to update whether we're only showing withdrawn enrollments
+        updateWithdrawnQuery: (withdrawn: boolean) => {
+          history.push({
+            search: stringify({
+              ...query,
+              month: defaultCtx.query.month,
+              withdrawn,
+            }),
+          });
+        },
       }}
```
Updating the query really means updating the history, since the `useEffect` at the top of his ctx will respond to those changes, updating the query state

vonbearshark commented on src/routes/children.ts:
```diff
@@ -23,37 +23,90 @@ childrenRouter.get(
+      forceArray: true,
+    }) as string[];
+    const showFundings = parseQueryString(req, 'showFundings') as string;
+    const metadata = await controller.getMetadata(
+      req.user,
+      organizationIds,
+      showFundings === 'true'
+    );
+
+    res.send(metadata);
+  })
+);
+
+childrenRouter.get(
+  '/missing-info',
+  passAsyncError(async (req, res) => {
+    const organizationIds = parseQueryString(req, 'organizationId', {
+      forceArray: true,
+    }) as string[];
+    const missingInfoChildren = await controller.getMissingInfoChildren(
+      req.user,
+      organizationIds
+    );
+
+    res.send(missingInfoChildren);
+  })
+);
+
+childrenRouter.get(
+  '/validation-errors',
+  passAsyncError(async (req, res) => {
+    const organizationIds = parseQueryString(req, 'organizationId', {
+      forceArray: true,
+    }) as string[];
+    const validationErrors = await controller.getErrorCounts(
+      req.user,
+      organizationIds
+    );
+
+    res.send(validationErrors);
```
separate routes rather than one `/` route with entangled param paths

vonbearshark commented on client/src/containers/Roster/hooks/usePaginatedChildData.tsx:
```diff
@@ -2,7 +2,7 @@ import { stringify } from 'query-string';
 import { useAuthenticatedSWRInfinite } from '../../../hooks/useAuthenticatedSWR';
```
No real changes here, but would like a place to talk about pagination. I am wondering how we want to handle the separation by `site`, specifically. Right now the client is handling the `site` filter. That _might_ be okay. But it does get trickier with pagination, I think. Like, what happens when we are on a `site`'s tab and we `load more` by age? what do we fetch exactly? Seems like there is some complexity there and interested in thoughts (@jlhogan in particular)

vonbearshark replied:
capturing form irl: mvp could easily be to just throw out data from site filtering pagination calls without sophistication around reconciling previously fetched/cached state

jlhogan replied:
heh i put a comment capturing same idea above. yes! we will get there! once this, and then better pagination, lands.

vonbearshark commented on client/src/containers/Roster/Roster.tsx:
```diff
@@ -198,60 +107,33 @@ const Roster: React.FC = () => {
-        ) : (
-          <RosterContent {...rosterContentProps} />
-        )
+  // Organization filtering happens on the server-side,
+  // but site filtering needs to happen in the client-side, if a
+  // site is requested
```
Worth talking about this. It would be nice to at least group by site on the backend? Might run up against pagination changes, though--see the other comment

jlhogan replied:
documenting what we discussed: site filtered vs unfilitered data does not need to be smartly re-used; can move this filter into the backend and re-request individual pages based on what roster the user is looking at. IMO only worth doing (moving site filter into db-level in server) once we've updated pagination to let the user do it.

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -134,109 +236,93 @@ export const getSiteCountMap = async (children: Child[]) => {
+  return fundingSpacesDisplay;
+};
+
+export const getErrorCounts = async (user: User, organizationIds: string[]) => {
```
TODO: SQL-ize. But wondering if that's even possible given how we manage validation errors? Aren't they meta information? Are they ever in the db?

JordanGuinn replied:
I think it'd be **really, really** annoying to generate equivalent SQL queries out of the processing behavior we have defined here 😕

JordanGuinn replied:
Also, I can't help but hear `SQL-ize` as [Sequelize](https://sequelize.org/) and get very sad at all the trouble TypeORM has given us 😅😭

bamader replied:
I do my best to not think about TypeORM at all.




Except...
Sometimes, late at night...
That's when the demons come... 👻

bamader replied:
But also yes I think it would be a really big frustration to try to turn our validation processing into SQL 😞

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -19,79 +80,116 @@ import { getCurrentEnrollment } from '../../utils/getCurrentEnrollment';
+      const start = activeMonth.startOf('month').format('YYYY-MM-DD');
+      const end = activeMonth.endOf('month').format('YYYY-MM-DD');
+
+      // TODO filter and only return matching enrollments
```
In the previous code we are only returning enrollments that match--is that actually necessary, though, as long as the conditions are satisfied?

vonbearshark replied:
Did an investigation with @bamader. It _seems_* like the only place we really use enrollments as returned from this fn is in the Roster view, in the `Enrollment Date`. However, this is actually kind of broken, as is. We use this `Enrollment Date` in past (`month`) filters. And that doesn't make sense, really. Children can have multiple enrollments in a past filter (especially once we expand it beyond just a month). Right now we just pick the "newest" one, but that is kind of arbitrary? My pitch would be _not_ to filter the enrollments for the past, and instead have this field always show the most recent enrollment date. We might want to update the language to  `Current Enrollment Date` or `Most Recent Enrollment Date` (to apply to the `withdrawn` filter as well, where "current" doesn't make sense).

*I still need to investigate. I am wondering if the other fields in the Roster view, like `Funding type`, `Space type`, and `Site` are derived from the most recent enrollment as well. I am thinking that they might, and that would kind of open a can of worms. Again, then they are _all_ broken/arbitrary for `month` and `withdrawn` filters right now, and my pitch would probably be to make it clear that _all_ of the fields are `Most Recent`

does that make sense? Thoughts? (cc @jlhogan maybe)

vonbearshark replied:
I guess on second thinking picking the most "current" enrollment that also satisfies the filter (`withdrawn` or `month`), is not that arbitrary?

vonbearshark replied:
(addressed this with the new sql-izing commit)

vonbearshark commented on src/utils/processChild.ts:
```diff
@@ -1,23 +1,10 @@
+import { Child, Enrollment } from '../entity';
```
Why do we need to process the Child entities? Should the processing be happening as we go, rather than as we are fetching the data?

vonbearshark replied:
If it is necessary then `TODO` I am still handling it in the `get.ts` refactor

JordanGuinn replied:
What do you mean by `happening as we go` exactly?

I know that all children displayed in the roster need to be processed (validated) so that we can display that little `Missing Info` icon for those rows, but once we start applying pagination to the roster queries, that'll only be necessary for the children within the page being queried for.

vonbearshark replied:
Ah. That is kind of frustrating but understandable. Right now we are doing

```
export const postProcessChild = async (child: Child) => {
  removeDeletedEntitiesFromChild(child);
  sortEntities(child);
  return validateObject(child);
};
```

Do we need to remove deleted entries or sort entries? Why isn't the former done elsewhere (wherever entries are deleted)? And is there a specific reason we do the latter here? I assume it's for convenience for the ui, but it might make sense elsewhere post refactor

JordanGuinn replied:
1. Agreed, sort should def happen at the db layer, because of course it should
2. Manually removing rows with `deleteDate` values because apparently TypeORM doesn't too well support implicit filtering of rows with such values.  We have [an outstanding question, in ticket form,](https://app.zenhub.com/workspaces/ece-dev-board-5ff506028d35f30012e0e937/issues/ctoec/data-collection/1174) to figure this out

vonbearshark replied:
double check, but I think I was able to add this to the sql-izing commit. essentially I left join _and_ do this sanitization when bringing in the relations

vonbearshark reviewed (commented):
`get.ts`, `children.ts`, `RosterContext.tsx`, and `Roster.tsx` have the big changes

vonbearshark commented on src/routes/children.ts:
```diff
@@ -23,37 +23,90 @@ childrenRouter.get(
+);
+
+childrenRouter.get(
+  '/validation-errors',
```
I don't love this extra call for error data. When implementing the pagination it is possible there is an initial request that has the first paginated payload _plus_ the error data. On one hand, we probably do want to have that error count updating as errors are resolved. Otoh that might not be happening now as we expect, anyway...

vonbearshark replied:
Following that thought: the existing functionality here is 1 broken due to edge cases in `active` v `withdrawn` 2 handled in the client. Handling in the client is cumbersome, but it does allow for "live" updating of the active errors. I am not sure how to best account for that with a backend-based system, since we need a client condition to trigger a recompute of the error state. We could do it 1 on initial pagination payload + 2 on submit attempt, but it is not ideal since the count won't update in the meantime

JordanGuinn replied:
Ugh, that's tough.  

I honestly don't think it would be an issue to be overly chatty with this API, if it didn't require pulling _all the children in the roster_ off row 😬 We wouldn't even get the benefit of piggybacking off the child SELECT when retrieving paged roster data either, cuz it'd require an entirely separate, unpaged query anyway.

JordanGuinn replied:
Certainly not applicable to this PR, but maybe it's time we start thinking about storing this ephemeral roster error metadata somewhere

vonbearshark replied:
Slight spoilers, but I sqlized the child data fetching (not the metadata calls). I realized while doing that they we can really _never_ sqlize the child validation stuff. Like, in multiple cases we need to grab all of the children and count al of their errors. That will end up being a bottleneck on some routes.

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -19,79 +80,116 @@ import { getCurrentEnrollment } from '../../utils/getCurrentEnrollment';
+  organizationIds?: string[],
+  selectParms?: {
     activeMonth?: Moment;
     skip?: number;
     take?: number;
```
Am I missing something? I thought we needed to `ORDER BY` to use these params/pagination?

vonbearshark replied:
I guess I mean: how are we getting away without it now at all? Like, shouldn't typeorm be pissed at least?

JordanGuinn replied:
Oh I get you - I honestly think it's just dumb luck.  I know that using `OFFSET` and `FETCH` without `ORDER BY` in SQL Server is generally a really bad idea (as there's no implicitly ordering that can be relied upon consistently), but I don't know it to be technically impossible.

Least it sure doesn't seem like it is here 😅

JordanGuinn reviewed (approved):
Okay, first pass at this PR complete.  Got a handful of questions/comments, but overall this is looking _real real_ good 🙏🏽 Also, I do think it addresses everything I had in my previous PR!  I can incorporate that testing work into validation here just to be extra sure, though

jlhogan commented on client/src/containers/Roster/Roster.tsx:
```diff
@@ -273,44 +155,55 @@ const Roster: React.FC = () => {
-          )}
-          <LoadingWrapper text="Loading your roster..." loading={loading}>
-            {rosterContent}
+          {!query.withdrawn && <RosterButtonsTable />}
```
this is not a change that you're making, but why are we hiding the buttons table in withdrawn view?

bamader replied:
Only thing I can come up with is that we didn't want users adding children with withdrawn enrollments from the create flow, but that doesn't seem right or like something we'd do

jlhogan commented on src/routes/children.ts:
```diff
@@ -23,37 +23,90 @@ childrenRouter.get(
     }) as Moment;
     const skip = parseQueryString(req, 'skip', { post: parseInt }) as number;
     const take = parseQueryString(req, 'take', { post: parseInt }) as number;
-    const siteMap = parseQueryString(req, 'siteMap');
```
yeah no it's bad. but at some point i was annoyed that no nice typescripty thing for this existed and wrote one. but overall i agree that it's weird and not actually very useful.

jlhogan reviewed (approved):
Overall this gets us into a much better state. In terms of validating, as I mentioned I think the best way I can understand to have HIGH CONFIDENCE that we didnt introduce regression would be to test with prod data against prod views.

Also happy to hammer at it in QA once we push it there! I have at least a bit more of an idea of what/how to manually validate so we don't feel like we're just flailing in the dark :)

bamader commented on client/src/containers/Roster/hooks/useIsSubmitted.tsx:
```diff
@@ -0,0 +1,25 @@
+  useEffect(() => {
+    if (!organizationId) return;
+
+    apiGet(`oec-report/${organizationId}`, accessToken).then((res) => {
+      setIsSubmitted(!!res?.submitted);
+    });
+  }, [organizationId, accessToken]);
```
How many users are there with access to multiple orgs? Since we don't show the post-submit version of the home page until _all_ of a user's orgs have submitted their data (which is represented by the same `oec-report` check), is that something we need to consider here?

vonbearshark replied:
ug. Is it? hm..

vonbearshark replied:
I am not sure I see how we were handling that case before

vonbearshark replied:
@jlhogan do we have a switch for multiorg users to only select one roster/org at a time? Talking with @bamader it might be a vague design sitch

vonbearshark replied:
Ah so we are good here. This is only used in the `Roster` view, and it is on a per-org basis 👍🏻

bamader commented on src/controllers/children/get.ts:
```diff
@@ -1,16 +1,77 @@
+ * and enrollments for those sites
+ * Accepts QueryBuilder extension function to manipulate the SELECT query.
+ */
+const findChildrenOpts = async ({
```
Typescript will yell at us if we ever leave an `await` that doesn't do anything (it flashes a warning like 'this await has no effect') so I think we tend to correctly use it? But also, I'm just paranoid about race conditions and synchronous work so I will await errything if I need to.

bamader commented on src/controllers/children/get.ts:
```diff
@@ -1,16 +1,77 @@
+ * Accepts QueryBuilder extension function to manipulate the SELECT query.
+ */
+const findChildrenOpts = async ({
+  user,
```
I think the last sentence is probably the most relevant analysis. I'd hate to put in all kinds of effort parameterizing something to accept different search queries when to me at least it's not awful to just invoke whichever of `find` vs `findOne` we need. Code laziness ftw (at least until it impacts other stuff--then burn it all 🔥 )

bamader commented on src/controllers/children/get.ts:
```diff
@@ -1,16 +1,77 @@
+}) => {
+  const readOrgIds = await getReadAccessibleOrgIds(user);
+  const orgIds = organizationIds?.length
+    ? intersection(readOrgIds, organizationIds)
```
I think a lot of the org filtering scattered around here is more of a pseudo-security thing (access control is probs a better phrase but I'm feeling dangerous today). Like, we want users who are associated with particular organizations to be the only ones actually making changes for those organizations. If a user doesn't have that `orgId` credential, then typeorm query won't return anything from `find` and their operation will have no effect.

vonbearshark replied:
Hm. We should satisfy that kind of security ish requirement with the `findChildrenOpts` fn, which only allows you to select from orgs you have access to, and it also defaults to all of your accessible orgs.

What about from a product perspective. How do we handle multi-org users?

Right now there is a nontrivial amount of frustrating code (especially in the frontend) around providing a mandatory org ID for all Roster queries. I suspect we should be able to eliminate that, and worst-case we could just allow for a specific org ID override for multi org users selecting a particular org. But I don't understand the app well enough to say for sure. @JordanGuinn @jlhogan maybe

JordanGuinn replied:
I haven't been able to confirm the user experience for multi-org users yet, but if there isn't a way for those types of users to be able to filter their roster contents by organization, then agreed, org filtering ought be an implicit permissions thing handled on the back-end and not based in any way on parameters provided by the client.

vonbearshark replied:
> if there isn't a way for those types of users to be able to filter their roster contents by organization

Wow yeah I have been operating under the assumption that this was necessary but it could be totally not? geez that would really simplify this code

vonbearshark replied:
capturing convo with @jlhogan 

> for multi-org users, we give them one org roster at a time. they can select which org by an additional UI nav component which is another level of tab nav
> so for the multi-org user we will pass an org param, and all other users we wont and they'll implicitly get "all" of their data but it will just be for one org

So this doesn't save us complexity here/on the backend, but it _should_ save us some frontend complexity where we are working to inject/force an org ID where we don't have to be all the time...

bamader reviewed (approved):
Wow that took some time and some staring to grok. I think I've wrapped my head around everything, though, and I _think_ everything looks good. Obviously, it'll be subject to copious amounts of validation, so once we as a team start getting on that train, I'll be happy to manually try to explode everything.

vonbearshark commented on client/src/containers/Home/PostSubmit.tsx:
```diff
@@ -34,33 +34,26 @@ export const PostSubmitHome: React.FC = () => {
-    }
   }, [accessToken]);
 
+  console.log('here');
```
Weird we don't warn about this? Or do we?

JordanGuinn replied:
We don't warn about this, unfortunately 😕 While we **do** have loggers set up for SQL queries and incoming HTTP requests, we **don't** currently have a dedicated logger for application usage, so there are a handful of places where we're using `console.log` meaningfully.

Maybe it's time we finally get an app logger set up, though 😁

vonbearshark commented on client/src/containers/Upload/ReviewMissingInfo.tsx:
```diff
@@ -11,7 +11,6 @@ import { BackButton } from '../../components/BackButton';
 import { EnrollmentColumnError } from '../../shared/payloads';
 import { getH1RefForTitle } from '../../utils/getH1RefForTitle';
 import { misingInfoTableColumns } from './missingInfoTableColumns';
-import { ReactComponent as Arrow } from '@ctoec/component-library/dist/assets/images/uswds/arrowRight.svg';
```
changes in this file unrelated

jlhogan commented:
@vonbearshark is this awaiting another round of reviews or any other interaction?

vonbearshark commented:
still working off feedback from yesterday and whack a mol-ing tests, but later today, I think

vonbearshark commented:
New commit just dropped!

Please re-review the changes (and thank you!)

Big diffs: 
- SQL-ized all child retrieval in `get.ts` 
    - including the processing (like sorting entries and removing soft deletes)
    - _not_ including validation error-based things or metadata calls
    - Needed to convert to querybuilder format. lmk what you think there
    - Not a typeorm fan!
- Updated the orgId and site query params' mgmt in the frontend
    - No longer force orgId injection throughout
    - mgmt now happens in roster ctx, again as source of truth (rather than individual components handing `history` directly

I will try and comment throughout the changes. I have all tests running locally, and I have been doing some misc manually QA. Next steps would be: finish code reviewing, then start on validation.

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -112,35 +134,25 @@ export const getActiveChildren = async (
+
+  const qb = await queryBuilderBuilder({ user, organizationIds });
+  qb.andWhere(
+    'enrollment.entry <= :end AND (enrollment.exit >= :start OR enrollment.exit IS NULL)',
```
Please confirm the thinking here. active for time = overlapping with that range at all, so if an enrollment starts before `end` of range and either ends after `start` of range or doesn't end at all.

bamader replied:
Would definitely appreciate other thoughts on these conditionals, cuz JJ and I came up with them during pairing, but these make sense to me. We ran through a couple of scenarios of when records should and shouldn't qualify for this filter and everything seemed to us to check out.

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -152,35 +164,24 @@ export const getWithdrawnChildren = async (
-  );
-  return children;
+  const qb = await queryBuilderBuilder({ user, organizationIds });
+  qb.andWhere('enrollment.exit IS NOT NULL');
```
Confirm the thinking here: withdrawn = children where all enrollments have been terminated

(not _exactly_ sure why this `where` works with the query, but it seems to?)

bamader replied:
Yeah this should be the right query, to me. `.exit` is the date a child left an enrollment, and it defaults to null for all enrollments that the child is presently in. So, it's only _not_ null if a child has left the enrollment, hence the withdrawn status.

jlhogan replied:
why do we still have a `getWithdrawnChildren` endpoint? Shouldn't this just be the time-filtered endpoint, but for start date = inifinity ago and enddate = yesterday ?

vonbearshark replied:
Yeah that makes sense to me!

vonbearshark replied:
Hm actually this might not align with my current thinking on the temporal filter. Right now it captures all enrollments that overlap with the range, as long as they start before the end of the range and either don't end or end after the start of the range. So it gets enrollments that started before and continued through, those that started in the middle and went on, and also those that both started and ended in the range. `-Infinity` to `yesterday` would capture children `withdrawn` today because they meet that last criteria, right? Since they have all their enrollments contained within a range.

jlhogan replied:
ahhh yes, so i have the range wrong then. the full withdrawn roster is start: yesterday, end: yesterday). So that every enrollment going back in time meets first criteria (start before end of range) and all enrollments that are withdrawn as of yesterday meet second criteria (end after start of range). yeah ? time math is hard, but i know this is doable and we defintiely don't need a withdrawn end point.

vonbearshark replied:
Ah yep, that tracks!

vonbearshark replied:
Hm. Thinking thought now. That yesterday,yesterday gets everything. And that does make sense, I think. As you said, all enrollments meet the first criteria. They also meet the second because they have an active enrollment, so that is still the first criteria: they started before yesterday and went on through...

vonbearshark replied:
We have made the time filter super broad (any enrollment that overlaps with it at all) and the withdrawn filter super specific (at least one active enrollment), so I am actually less confident we can rm the withdrawn filter

jlhogan replied:
So, my understanding of the desire to move to this model of time-based filtering was in part explicitly to serve both of these views and remove the "withdrawn" filter. there is not a product need to give users the ability to select random time slices; providers will likely always just want a month roster. 

But now that i'm actually applying my brain to it, i see that you are right because we need to flip off that "exit is after start OR exit is null"

perhaps a thing to do here would be:
if start and end params provided:
- entry is before end AND (exit is after start OR exit is null)
if only end param provided (withdrawn roster!):
- exit is before end
if only start param provided (is this needed? current roster?)
- entry is after start

jlhogan replied:
discussed IRL -- withdrawn =/= normal time filtered active roster!!!! keep separate route and investigate if withdrawn as a view/route/whatever is actually a need

vonbearshark replied:
Ug so this is _not_ correct as is. This was my suspicion originally that I couldn't articulate: as is this filters down the enrollments that have exits, per child. However, we really want those children where _all_ the enrollments are not active.

You would think it would be straightforward to even just `COUNT` the enrollments here or something, but after an hour of wrestling with `typeorm` I am pretty burnt on trying. Likewise also tried to augment the active fn/query as @jlhogan suggested, but it ran into the same thing, essentially: I could create conditions that filtered to only withdrawn enrollments, but couldn't return only the children with _only_ withdrawn enrollments.

Open to thoughts here

vonbearshark replied:
Also capturing an irl potential to just remove the withdrawn filter from @jlhogan : that would be trickier than it seems. As is the active and withdrawn filters are compliments. Without the withdrawn filter we would have no way of accessing those children. And they currently contribute to counts throughout the app. So I do think it is at least most straightforward to keep the filter itself for now

JordanGuinn replied:
Definitely don't have any good suggestions with query builder, but could we potentially use `NOT EXISTS` here?  Like only select children where there aren't enrollment rows without an exit date.  Forgive my pseudo-SQL, but something like:

```
SELECT *  
FROM Child as c  
WHERE NOT EXISTS  
(SELECT *   
    FROM Enrollment AS e
    WHERE e.childId = c.id 
    AND e.exit IS NULL) ;
```

vonbearshark replied:
It does seem like any SQL-y solution would require a separate subqery on `Enrollment`s, which is not ideal (or theoretically necessary), and makes it a little tough to reuse with the current querybuilder builder. I was trying to avoid it, but I do think I could make it work like that, and I think it's preferable to post processing in memory.

JordanGuinn replied:
Agreed on both.  If there _is_ a SQL solution here that doesn't require another subquery on `Enrollment`, I'm not seeing it.

vonbearshark replied:
Did something of this effect in last commit. Feel free to take a gander. Can confirm it seems to work

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -1,76 +1,100 @@
+  });
 
-  return opts;
+  return qb;
```
Maybe I have been staring at it too long, but is this so bad? It is tedious, but at least it's clear what is happening?

jlhogan replied:
no i think it's definitely the better way to do it -- initially was trying to stick with the niceness of the "find" operations, but they quickly became way less "nice". All the TypeORM docs indicate that this is the intended/expected usage approach for more complex querying. Overall very agree tedious but clear

bamader reviewed (approved):
This looks good to me. I like the SQL changes that were made, as well as some improvements to the query parsing and general simplification. Pending some thorough manual validations, this looks great!

vonbearshark commented on client/src/contexts/RosterContext/RosterContext.tsx:
```diff
@@ -43,122 +41,28 @@ const defaultCtx = {
+  const { error, ...childData } = usePaginatedChildData(query);
   if (error) throw error;
 
-  /**
-   * useSWR provides a `mutate` callback to update cache values, but I cannot
-   * get it to work for the inifite flavor of the hook. This is a workaround
-   * that directly mutates the cache that underlies the SWR stuff without
-   * any of the other logic baked in to the mutate func (logic around if the
-   * changes should actually be applied, I guess? because they're not)
-   * @param updatedChild
-   */
```
Was able to [fix](https://github.com/ctoec/data-collection/pull/1253/commits/8908d981b6abbe0b75115643d74c82b605a7355b#diff-ff0631363477e3e64ca2787bda67ba6b16fe865a67d3fbc52ccb90a033a7289dR84) this I think

jlhogan replied:
:joy:

vonbearshark commented on client/src/hooks/useAuthenticatedSWR.ts:
```diff
@@ -20,7 +20,7 @@ export function useAuthenticatedSWR<T>(
   const { accessToken } = useContext(AuthenticationContext);
   return useSWR(
-    !path ? null : [path, accessToken],
+    !path || !accessToken ? null : [path, accessToken],
```
correct me if I am wrong here, but this is what we want for "authenticated"  calls in particular?

jlhogan replied:
yessssssssssssssssss

vonbearshark commented on client/src/hooks/useAuthenticatedSWR.ts:
```diff
@@ -39,7 +39,7 @@ export function useAuthenticatedSWRInfinite<T>(
   const getKey = (index: number, prevData: T | null) => {
     const path = getPath(index, prevData);
-    const key = !path ? null : [path, accessToken];
+    const key = !path || !accessToken ? null : [path, accessToken];
```
again correct me if I am off base

vonbearshark commented on client/src/containers/Roster/hooks/usePaginatedChildData.tsx:
```diff
@@ -31,15 +35,12 @@ export const usePaginatedChildData = ({
       return null;
     }
 
-    // Paginated api query (but only once we have organizationId param)
```
Doesn't seem to be a reason for this

jlhogan replied:
`useSWR` doesn't make the network request if this is null. Previously if we didn't have org id, we didn't want to make the network request. Now, org-id isn't required to make the API request. 

However, for a multi-org user, will this api call be kicked off to grab data for all orgs without the org filter in some state prior to that "activeOrgId" being set? ((^^thing to validate))

vonbearshark replied:
Added the fix in for this. Essentially a way simplified version of the `useEffect` we had previously. We just inject the orgId for multiorg users on first load, but without the `useEffect`. Due to our refactor, that is fine

vonbearshark replied:
When we end up at `/roster` from another page (via `history.push`, or updating the url in browser, refresh, etc.), it is injected on that first load. And now with the refactor that query is only updated from the fns exposed by the ctx. And those fns all preserve the pre-existing query, only updating a piece at a time. Finally, the fn that we have to update the query's orgId requires a new orgId.

So in short there is no way to "unset" the orgId once it is added, and it's added right away.

jlhogan commented on client/src/containers/Roster/Roster.tsx:
```diff
@@ -132,8 +131,8 @@ const Roster: React.FC = () => {
           location={
-            query.month || query.withdrawn
-              ? `/roster?organization=${query.organizationId}`
+            (query.month || query.withdrawn) && activeOrgId
```
i know this isn't actually a diff, but i dont understand this conditional -- go back to roster if we were looking at a filtered roster, otherwise go back to home page?

vonbearshark replied:
I _think_ we do want the `Back` button the Roster page in a filtered view to go back to an unfiltered roster, which this effectively does. The `orgId` bit is a little wonky, because the logic as written only does that for a selected orgId, like in multiorg users. So that should probably change regardless

vonbearshark replied:
(updated this. think it was actually broken and now fixed)

jlhogan commented on client/src/contexts/RosterContext/RosterContext.tsx:
```diff
@@ -43,122 +54,41 @@ const defaultCtx = {
   const history = useHistory();
   const query: RosterQueryParams = parse(history.location.search);
 
-  // Parse and sanitize query params, update if missing (i.e. initial load) or invalid
```
:wave:

vonbearshark replied:
well based on your investigation I am maybe not so sure...

vonbearshark replied:
psych, yep we don't need this. we just need to update the `query` object on first load like normal, without `useEffect` (it is not really state, anyway)

jlhogan reviewed (approved):
code review approval -- manual validation happening now!

jlhogan commented:
manual validation:
- [ ] roster filtered for past month only includes records for children with active enrollment during that month
- [ ] roster filtered for past month displays current enrollment at time of filter

buggy tings:
- roster filter always says "Invalid date"
![invalid_date](https://user-images.githubusercontent.com/54035880/113710742-17a58280-96b2-11eb-99e6-88caa8359367.png)
- upon opening filter modal, immediately presented with error state for empty default filter value: 
![date_error](https://user-images.githubusercontent.com/54035880/113710914-4facc580-96b2-11eb-8d83-eaee27f5a360.png)
- not really within scope here, and maybe better addressed by the work @nickbristow is currently doing, but confusing that the example date for date picker is in "M D YYYY" format when the date input is only asking for "M YYYY"

jlhogan commented on src/controllers/children/get.ts:
```diff
@@ -1,16 +1,101 @@
+  const qb = await getManager()
+    .createQueryBuilder(Child, 'Child')
+    // Restrict via org access
+    .where('Child.organizationId IN (:...orgIds)', {
```
does there need to be a whereNotDeleted for the child entity?

vonbearshark replied:
It seems like the [original](https://github.com/ctoec/data-collection/pull/1253/files#diff-8599112c8a94c9f008727779a95f5e7d9f89aa28f5241c0fcbef61489a6e75d8L21-L32) was only checking for the existence of the child itself, which should happen on the join anyway, right? But I could add it if it's that prop exists on the child potentially

jlhogan replied:
looked into a bit more, createQueryBuilder automatically includes the deleted check for the top-level entity!

vonbearshark commented:
re: bugs

> roster filter always says "Invalid date"
> upon opening filter modal, immediately presented with error state for empty default filter value:
I actually can't repro these...

Happy to keep at it but between these and the prompt format work I might say that if they are flaky bugs they can be fast-followed with either @nickbristow 's new date picker work, or the arbitrary time range ticket I was planning on taking right after (and should be short work). probably a lot of overlap in those follow tix..

vonbearshark commented:
Updating the latest commit (since the changes were relevant) and force pushing. Added context for the multiorg situation in comments.

Re: the weird auth bug: my [fix](https://github.com/ctoec/data-collection/pull/1253/files?file-filters%5B%5D=.tsx#diff-ae2208b31ad9bf2c689e61b320d244271531757d3d400a55ae4dcff806dced28R82) is a little unsatisfying. We were `throw`ing an error rather than just logging it. 

1. I think that is the preferred flow regardless? Correct me if I am wrong. There are other solutions now that I know what's going on here, but I generally don't think we should be `throw`ing like this in the client based on server responses?
2. The actual issue in this case is related to the `handleJWTError` flow ( in [`client/src/utils/handleJWTError`](https://github.com/ctoec/data-collection/blob/base/client/src/utils/handleJWTError.ts)), and it actually seems like a perfectly acceptable course of action that was unnecessarily blocking here with the `throw`. Logging it allows the flow to complete and the client to reauth without any noticeable issues at all. Again, I can see about updating this flow so we can keep `throw`ing, but idt that's what we want

jlhogan reviewed (approved):
Can confirm: 
- [x] cyclical logout loop is resolved
- [x] multi-org roster queries are filtered by org, and so only appropriate data is displayed in first roster all sites section

bamader commented:
General roster and flow functionality tests:
- [] Create record flow can add a completed child
- [] Create record saves progress after breaking at each step without crashing roster (new caching)
- [] Batch edit can make and save all edits for a record
- [] Batch edit can do some edits then refresh/return to roster; changes should be saved, nothing should crash
- [] Delete record gets rid of the record from roster (including after refresh) and updates the number at the top of the page
- [] Edit record: each of the five sections can make and save edits to children; these persist through saving via caching and selecting options does not cause non-responsive behavior from the app
- [] Withdraw record removes the record from the active page and updates numbers at the top of page (persists through refresh)
- [] Can time filter for a record immediately after withdrawing it
- [] Can change sites on roster page and see correct children (active + withdrawn); appropriate params are attached to / updated in query
- [] Can change orgs on roster page and see correct children (active + withdrawn)
- [] Submitted success alert persists on roster after submitting data (including after refresh)
- [] Number of errors claimed at the top of roster page matches real number of validation errors
- [] Can only submit data if an org-level user
- [] Roster returns to normal after clearing the time filter
- [] Preview of an upload is correct and proceeds to the roster with the advertised children

Test cases for specific retrieval / roster bugs we've had in the past:
- [] Adding income determinations doesn't cause render errors, and the new dets are retrieved successfully
- [] The correct income det UI displays for children with and without dets
- [] Not disclosed determinations display correctly
- [] The correct enrollment and funding UIs display for children with no enrollments, children with unfunded enrollments, and children with at least one funded enrollment
- [] Reporting periods are parsed forgivingly and without validation errors when uploading sheets
- [] Funding space times are still mapped equivalently and retrieved properly
- [] Pre and post submit pages show correct sections (sites, funding maps) for users at site/org perm levels

vonbearshark commented:
New commit addressing the bugs @bamader  found. Retriggering tests and deploys and should be ready for revalidation!

Fixed:
- Batch edit banner issues (as well as unrelated bug found in batch edit process)
-  Can only submit data if there's no errors failure
- "Redetermine" bug
- Alert persists on the page after refresh and going through Edit Record work
- Preview -> Roster slow/wonkiness

Couldn't repro:
- Saving a child after inputting just section 1 in create record, then navigating back to roster, then to the create flow crashes the app
-  Number of errors claimed at the top of roster page matches real number of validation errors

Questions:
-  can navigate between pages without hanging or flashing UI from the app
    - This was unchecked (and its parent), is there an issue here you observed @bamader?

Notes:
- Some of the repro issues were tested after some fixes, so it is possible I just fixed them. Also possible I don't understand the context, though.
    - I fixed a pretty gnarly race condition that effected basically the whole Edit experience (and maybe more?) that could easily cloud my repro and affect related behavior
- Also realized the Error banner was wrong (in addition to not showing), so fixed that as well
- I fixed the Preview -> Roster loading lag with `revalidate`, which is [apparently](https://github.com/vercel/swr/discussions/812#discussioncomment-203624) being deprecated. You are supposed to be able to replicate the same functionality with `mutate`, but I haven't had any luck. Left a `TODO`.
- Unrelated, but after poking found that the [withdrawn filter](https://github.com/ctoec/data-collection/pull/1253#discussion_r610339924) is not working as intended 🙄 ...

bamader commented:
Ooooh lots of new things for me to test 👀 . Will start validating again in an hour or so after I clean up some other stuff. After I check that, I'll see if I can brainstorm any fixes for the Withdrawn filter.

My first thought is a weird hybrid combination of the current work and the old functionality of processing in post. Right now, the code finds all enrollments that have an exit, then gets the children associated with those enrollments. This should, by definition be a set that fully contains children for whom _all_ enrollments are inactive. So what we can do is use a modified version of the current query _without_ any of the joins on nested relations (no families, income dets, etc.), which will obtain a list of child IDs corresponding to children that have at least one inactive enrollment. From there, we make  second query that retrieves _all_ information about each of these children, including the nested relations (we can use the SQL join syntax for this). Then, we do a post-processing `childrenWithAtLeastOneExitedEnrollment.filter( (c) => c.enrollments.all( (e) => e.exit ) )`, which would leave us the list of children for whom all enrollments are inactive. While this may be somewhat inelegant because it uses two queries to accomplish what we _should_ be able to write as one query (curse you typeorm), I think there are two reasons this approach might be okay:

1. JJ has already spent an hour+ trying to get this to work and I'm sure his sanity is being eroded by typeorm idiosyncracies 😕 
2. The first thing we query is not only a containing set for children with all inactive enrollments, but is also still _much_ smaller than a list of all children overall, so maybe the drop in performance won't be a super big deal if we have to make a second query.

Just my two cents on getting this put out sooner so JJ can take some well-deserved time on non query things! I'll get working more on validating this soon.

vonbearshark commented:
Yeah I generally agree that this is not blocking because at any point I could post-process my way out of this. If anything more sophisticated comes to anyone, please share. In the meantime I will be trying to make it work the Right™️ way. But whenever validation is successful I know I have a non optimal fix that can be validated quickly in isolation.

Also all e2e tests are passing 🎉 , and the only integration test failure is due to `withdrawn` count being off.

JordanGuinn commented on client/src/containers/Home/PostSubmit.tsx:
```diff
@@ -34,31 +34,22 @@ export const PostSubmitHome: React.FC = () => {
-      .then((res) => setSiteCountDisplay(res.siteCountMap))
+        // Also determine the funding spaces map for the organization, if
+        // the user has the permissions that enable this
+        if (fundingSpacesMap) {
```
nitty nit: should this conditional maybe be `(showFundings && fundingSpacesMap)`?  Just to confirm that we did in fact ask for it, in addition to it being provided.

vonbearshark replied:
yeah true

JordanGuinn reviewed (approved):
Second review pass complete - still looks good to me 👍🏽👍🏽👍🏽

bamader commented:
Posting a fresh testing suite here so I can work off a clean slate:

General roster and flow functionality tests:
- [x] Create record flow can add a completed child
- [ ] Create record saves progress after breaking at each step without crashing roster (new caching)
  - Saving a child after inputting just section 1 in create record, then navigating back to roster, then to the create flow crashes the app
- [ ] Batch edit can make and save all edits for a record
  - unable to test because the missing info banner disappeared
- [ ] Batch edit can do some edits then refresh/return to roster; changes should be saved, nothing should crash
  - unable to test because the missing info banner disappeared
- [x] Delete record gets rid of the record from roster (including after refresh) and updates the number at the top of the page
- [ ] Edit record: each of the five sections can make and save edits to children; these persist through saving via caching and selecting options does not cause non-responsive behavior from the app
  - [x] edits correctly save
  - [ ] can navigate between pages without hanging or flashing UI from the app
- [x] Withdraw record removes the record from the active page and updates numbers at the top of page (persists through refresh)
- [x] Can time filter for a record immediately after withdrawing it
- [x] Can change sites on roster page and see correct children (active + withdrawn); appropriate params are attached to / updated in query
- [x] Can change orgs on roster page and see correct children (active + withdrawn)
- [x] Submitted success alert persists on roster after submitting data (including after refresh)
- [ ] Upload success alert shows on roster and disappears after other actions are performed
  - Alert persists on the page after refresh and going through Edit Record work
- [ ] Number of errors claimed at the top of roster page matches real number of validation errors
  - this isn't working for me; I can make the missing info banner alert at the top of the page disappear through a combination of uploading, refreshing, going to edit record and making some changes, saving, then refreshing again
- [x] Can only submit data if an org-level user
- [ ] Can only submit data if there's no errors
  - this flat out doesn't work for me; I submitted with tons of errors and it let it go through
- [x] Roster returns to normal after clearing the time filter
- [ ] Preview of an upload is correct and proceeds to the roster with the advertised children
  - Preview is correct; after proceeding to upload, changes are not reflected in roster until refresh is done; even changing pages and then going back to roster doesn't fix this)
  - Also, if you try to upload a second file after uploading a first without refreshing in-between, the roster hangs and gets stuck in the loading state (I saw sticks of 1, 3, and 5+ minutes)

Test cases for specific retrieval / roster bugs we've had in the past:
- [ ] Adding income determinations doesn't cause render errors, and the new dets are retrieved successfully
  - Pressing "Redetermine" to add a new det overwrites the current income det on file; the new information is saved, the old det disappears 
- [x] The correct income det UI displays for children with and without dets
- [x] Not disclosed determinations display correctly
- [x] The correct enrollment and funding UIs display for children with no enrollments, children with unfunded enrollments, and children with at least one funded enrollment
- [x] Reporting periods are parsed forgivingly and without validation errors when uploading sheets
- [x] Funding space times are still mapped equivalently and retrieved properly
- [x] Pre and post submit pages show correct sections (sites, funding maps) for users at site/org perm levels

vonbearshark commented on src/controllers/children/get.ts:
```diff
@@ -19,79 +104,103 @@ import { getCurrentEnrollment } from '../../utils/getCurrentEnrollment';
-  return children;
+  const qb = await queryBuilderBuilder({ user, organizationIds });
+  // Subquery: find all active enrollments,
+  qb.andWhere((qb) => {
+    const subQuery = qb
+      .subQuery()
+      .select('e.childId')
+      .from(Enrollment, 'e')
+      .where('e.childId = Child.id AND e.exit IS NULL')
+      .getQuery();
+    return 'NOT Child.id IN ' + subQuery;
+  });
```
Worth a 👀

bamader replied:
Today I learned that subqueries are a thing 🤯 . This definitely looks right to me. Will incorporate into my validations.

JordanGuinn replied:
Confirmed - looks correct to me too

JordanGuinn commented on src/controllers/children/get.ts:
```diff
@@ -1,16 +1,101 @@
+ * If user has site-level access, also applies a site filter to only return children
+ * and enrollments for those sites. It also sorts enrollment and financial information.
+ */
+const queryBuilderBuilder = async ({
```
+100000 for the accurately named `queryBuilderBuilder` 😎

vonbearshark replied:
👷👷

JordanGuinn commented:
## Final Test Pass

### Upload
- [ ] Confirm I can upload the complete dev file successfully
- [ ] Confirm I can save said file to my roster successfully
- [ ] Confirm I can upload the dev file with partially missing fields successfully
- [ ] Confirm I can save said file to my roster successfully

## Roster
- [ ] Confirm the following within the normal roster view
   - [ ] Children without any enrollments are included
   - [ ] Children with an enrollment without an exit date are included
- [ ] Confirm the following within the month filtered roster view
   - [ ] Children with enrollments that ended prior to/during the filtered month are included
   - [ ] **No** children with enrollments without an exit date are included
- [ ] Confirm the following within the Withdrawn roster view
   - [ ] Children without any enrollments with an exit date are included
   - [ ] No children with an enrollment without an exit date are included
   - [ ] The sites tabs properly filters children by site

Create/Update
- [ ] Confirm I can create a new child without issue
- [ ] Confirm I can save the newly created child to my roster
- [ ] Confirm I can see said child within my roster
- [ ] Confirm I can batch edit children
- [ ] Confirm I can save the batch edits without issue
- [ ] Confirm I can see the edits reflected in my roster
